sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java

```diff
@@ -500,21 +502,28 @@ public void processElement(ProcessContext c) throws Exception {
                 ParDo.of(new WriteBundles<>(writeOperationView))
                     .withSideInputs(writeOperationView));
       } else {
-        PCollection<KV<Integer, T>> sharded;
         if (computeNumShards != null) {
           numShardsView = input.apply(computeNumShards);
-          sharded = input
+          ImmutableList.Builder<PCollectionView<?>> sideInputs =
+              ImmutableList.<PCollectionView<?>>builder()
+                  .add(writeOperationView)
+                  .add(numShardsView);
+          results  = input
               .apply("ApplyShardLabel", ParDo.of(
-                  new ApplyShardingKey<T>(numShardsView, null)).withSideInputs(numShardsView));
+                  new ApplyShardingKey<T>(numShardsView, null)).withSideInputs(numShardsView))
+              .apply("GroupIntoShards", GroupByKey.<Integer, T>create())
+              .apply("WriteShardedBundles",
+                  ParDo.of(new WriteShardedBundles<>(writeOperationView, numShardsView))
+                      .withSideInputs(sideInputs.build()));
```

Comment:
`.withSideInputs` accepts varags, if you'd like to save a few lines. The variable isn't reused, is it?